events/listener.go

```diff
@@ -0,0 +1,675 @@
+)
+
+// EventHandler defines the type for custom event handlers
+type EventHandler func(event interface{})
+
+// ListenerHandle is a handle returned when registering an event handler,
+// which can be used to stop listening to that event
+type ListenerHandle interface {
+	Unsubscribe()
+}
+
+// subscriptionInfo holds information about a subscription
+type subscriptionInfo struct {
+	subID       string
+	sub         event.Subscription
+	contract    common.Address
+	closeChan   func()
+	handler     EventHandler
+	unsubscribe context.CancelFunc // Used to cancel the resubscription
+}
+
+// EventListener listens for events from CMAccount and BookingToken contracts
+type EventListener struct {
+	client        *ethclient.Client
+	logger        *zap.SugaredLogger
+	mu            sync.Mutex
+	cmAccounts    map[common.Address]*cmaccount.Cmaccount
+	btContracts   map[common.Address]*bookingtoken.Bookingtoken
+	subscriptions map[string]*subscriptionInfo // Keyed by unique IDs
+}
+
+// NewEventListener creates a new EventListener instance
+func NewEventListener(client *ethclient.Client, logger *zap.SugaredLogger) *EventListener {
+	return &EventListener{
+		client:        client,
+		logger:        logger,
+		cmAccounts:    make(map[common.Address]*cmaccount.Cmaccount),
+		btContracts:   make(map[common.Address]*bookingtoken.Bookingtoken),
+		subscriptions: make(map[string]*subscriptionInfo),
+	}
+}
+
+// getOrCreateCMAccount gets or creates a CMAccount instance
+func (el *EventListener) getOrCreateCMAccount(addr common.Address) (*cmaccount.Cmaccount, error) {
+	if cm, exists := el.cmAccounts[addr]; exists {
+		return cm, nil
+	}
+	cm, err := cmaccount.NewCmaccount(addr, el.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create CMAccount instance: %w", err)
+	}
+	el.cmAccounts[addr] = cm
+	return cm, nil
+}
+
+// getOrCreateBookingToken gets or creates a BookingToken instance
+func (el *EventListener) getOrCreateBookingToken(addr common.Address) (*bookingtoken.Bookingtoken, error) {
+	if bt, exists := el.btContracts[addr]; exists {
+		return bt, nil
+	}
+	bt, err := bookingtoken.NewBookingtoken(addr, el.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create BookingToken instance: %w", err)
+	}
+	el.btContracts[addr] = bt
+	return bt, nil
+}
+
+// unsubscribe removes a subscription from the subscriptions map and closes the event channel
+func (el *EventListener) unsubscribe(subID string) {
+	el.mu.Lock()
+	defer el.mu.Unlock()
+	if subInfo, exists := el.subscriptions[subID]; exists {
+		// Cancel the context to stop resubscription
+		subInfo.unsubscribe()
+		// Unsubscribe from the event
+		if subInfo.sub != nil {
+			subInfo.sub.Unsubscribe()
+		}
+		// Close the event channel
+		if subInfo.closeChan != nil {
+			subInfo.closeChan()
+		}
+		delete(el.subscriptions, subID)
+		el.logger.Debugf("Unsubscribed subID %s", subID)
+	}
+}
+
+// listenerHandle is a handle to stop event listeners
+type listenerHandle struct {
```

Comment:
I'd suggest to move type and method below EventListener members for better grouping/readability

Reply:
also, maybe instead of struct you can define func() type? like

```go

type listenerHandle func()

func (l listnerHandle) Unsubscribe() {
	l()
}
```
also, why do you need type at all? can't you just use func() ?

Reply:
It's more explicit and self-descriptive with an `Unsubscribe()` method and possibility to extend the handle later. But, yeah sure, we cat do that when we need it. Switched to returning only a `func()`.